platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/inet/TcpIncomingConnector.java

```diff
@@ -109,13 +117,34 @@ public void run() {
                         InetSocketAddress remoteSocketAddress = (InetSocketAddress) socket.socket().getRemoteSocketAddress();
                         InetAddress remoteInetAddress = remoteSocketAddress.getAddress();
                         if (!allowRemote && !addressFactory.isCommunicationAddress(remoteInetAddress)) {
-                            LOGGER.error("Cannot accept connection from remote address {}.", remoteInetAddress);
+                            LOGGER.error("Cannot accept connection from remote address {}.", remoteSocketAddress);
                             socket.close();
                             continue;
                         }
-                        LOGGER.debug("Accepted connection from {} to {}.", socket.socket().getRemoteSocketAddress(), socket.socket().getLocalSocketAddress());
                         try {
                             SocketBlockingUtil.configureNonblocking(socket);
+                            ByteBuffer buffer = ByteBuffer.allocate(CONNECTION_PREAMBLE.length);
+                            long timeout = System.nanoTime() + TimeUnit.SECONDS.toNanos(acceptTimeoutSeconds);
+                            while (buffer.hasRemaining() && System.nanoTime() < timeout) {
+                                int read = socket.read(buffer);
+                                if (read == -1) {
+                                    break;
+                                }
+                                if (read == 0) {
+                                    Thread.sleep(10);
+                                }
+                            }
+                            if (!Arrays.equals(buffer.array(), CONNECTION_PREAMBLE)) {
+                                throw new IOException("Did not receive connection preamble within " + acceptTimeoutSeconds + "s");
+                            }
+                        } catch (IOException e) {
+                            LOGGER.error("Failed connection handshake with {}.", remoteSocketAddress, e);
+                            socket.close();
+                            continue;
+                        }
+
+                        LOGGER.debug("Accepted connection from {} to {}.", socket.socket().getRemoteSocketAddress(), socket.socket().getLocalSocketAddress());
+                        try {
                             action.execute(new SocketConnectCompletion(socket));
                         } catch (Throwable t) {
                             socket.close();
```

Comment:
❓ 

Not introduced via this PR, but I noticed we only close the socket in case of failure. what ensures the socket is closed in case of success?

Reply:
In case of success, the caller who provided the `action` is responsible for closing the connection when they are done talking to the other side.